purchases/src/main/kotlin/com/revenuecat/purchases/CustomerInfoRetriever.kt

```diff
@@ -0,0 +1,141 @@
+package com.revenuecat.purchases
+
+import android.os.Handler
+import android.os.Looper
+import android.util.Log
+import com.revenuecat.purchases.common.Backend
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.caching.DeviceCache
+import com.revenuecat.purchases.common.debugLog
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
+import com.revenuecat.purchases.strings.CustomerInfoStrings
+
+internal class CustomerInfoRetriever(
+    private val deviceCache: DeviceCache,
+    private val backend: Backend,
+    private val handler: Handler = Handler(Looper.getMainLooper())
+) {
+
+    fun retrieveCustomerInfo(
+        appUserID: String,
+        fetchPolicy: CacheFetchPolicy,
+        appInBackground: Boolean,
+        cacheCustomerInfoAfterFetchCallback: (CustomerInfo) -> Unit,
```

Comment:
This is the main problem I had after extracting the logic to this class from the `Purchases` object. After we fetch customer info from the backend, we store the new value in the cache and we send the updated value to the listeners if it has changed. Both operations are `synchronized` on the `purchases` object, so I thought moving them from there could cause concurrency problems.

Since this method is always called from `Purchases`, I added this callback to perform those 2 operations. Originally, I intended to use the already existing `ReceiveCustomerInfoCallback` instead of this new callback, but whether we execute the operations or not depends on if the customer info comes from cache or from the backend and we don't have that information on the callback. 

This solution seems okish considering it's internal but lmk what you think.

Reply:
I see your point...

Are we caching or doing anything else with the customer info in `Purchases` though? If we are not maybe it's fine to remove this callback and move the `cacheCustomerInfoAndSendToDelegateIfChanged` to this class and synchronize on it.

If not, I think we don't have any other solution, unless we create a private monitor in `Purchases` (instead of synchronizing on `Purchases`'s monitor) and pass it to this class?

Reply:
Also, I was a bit confused by the name of the callback, maybe we could name it `onSuccessfulFetchCallback`? That way the name doesn't dictate what the implementation of the callback is supposed to do (caching the customer info).

Reply:
`Purchases` also caches the customer info from some other sources, like when we restore purchases, we sync purchases with the backend or post a purchase to the backend. We could rename the `CustomerInfoRetriever` to `CustomerInfoHelper` and have all those cache pathways go through this class.

There is also the listener that is synchronized on the purchases object, and that was harder than the caching problem I would say... One solution I thought of is having the `Purchases` listener act as a proxy but having the actual listener live on the `CustomerInfoRetriever`/`CustomerInfoHelper` but I wasn't sure of this solution either...

Going to give it another thought but lmk if you would prefer something like that.